flink-table-store-core/src/main/java/org/apache/flink/table/store/file/mergetree/compact/AggregationMergeFunction.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.store.file.mergetree.compact;
+
+import org.apache.flink.table.data.GenericRowData;
+import org.apache.flink.table.data.RowData;
+
+import javax.annotation.Nullable;
+
+/**
+ * A {@link MergeFunction} where key is primary key (unique) and value is the partial record, update
+ * non-null fields on merge.
+ */
+public class AggregationMergeFunction implements MergeFunction {
+
+    private static final long serialVersionUID = 1L;
+
+    private final RowData.FieldGetter[] getters;
+
+    private transient GenericRowData row;
+
+    public AggregationMergeFunction(RowData.FieldGetter[] getters) {
+        this.getters = getters;
+    }
+
+    @Override
+    public void reset() {
+        this.row = new GenericRowData(getters.length);
+    }
+
+    @Override
+    public void add(RowData value) {
+        for (int i = 0; i < getters.length; i++)
+        {
+            Object currentField = getters[i].getFieldOrNull(value);
+            Object oldValue = row.getField(i);
+            Object result = sum(oldValue, currentField);
+            if (result != null)
+            {
+                row.setField(i, result);
+            }
+        }
+    }
+
+    private Object sum(Object oldValue, Object currentField) {
+        if (currentField == null)
+        {
+            return null;
+        }
+        if (oldValue == null)
+        {
+            return currentField;
+        }
+        if (oldValue instanceof Integer && currentField instanceof Integer)
+        {
+            return Integer.sum((Integer) oldValue, (Integer) currentField);
+        }
+        else if (oldValue instanceof Long && currentField instanceof Long)
+        {
+            return Long.sum((Long) oldValue, (Long) currentField);
+        }
```

Comment:
These `instanceof` checks will be performed on a per-record bases and have a big impact on performance. That's why we use `RowData.FieldGetter` instead of `instanceof` to fetch columns out of a row.

A better approach is to create an `Aggregator` class just like `RowData.FieldGetter` and in the constructor of `AggregationMergeFunction` we choose `Aggregator` for each column according to their types. So on a per-record path we can use the `Aggregator` directly without these checks.

An even better approach is to use Flink's existing code generation systems but that might be too complex for your first contribution, considering its usage and the fact that internal system of Flink table planner is now awkward to call from the outside due to the scala-free change in Flink 1.15. If you're interested see `AggWithoutKeysCodeGenerator` in Flink.

Reply:
should it be used
org.apache.flink.api.java.aggregation.AggregationFunction to handle aggregation? This is a convenient way to deal with different field types, but it also brings some problems. It is difficult for me to deal with delete/update RowData#getRowKind. How to solve it?
1. Modify the original interface org.apache.flink.api.java.aggregation.AggregationFunction to add delete and update functions
2. Implement the Aggregator class by yourself, manually switch the class of different fields and implement it?
Is there any other solution
For the type processing of object/class, I guess flink/flink-table-store already has useful tool classes to reduce development workload (such as org.apache.flink.api.java.aggregation.AggregationFunction, etc.), you can provide guide?

Reply:
I searched for a while and sadly I didn't find any class that can be reused. Many of the aggregate functions in the planner module are bound to code generation which can't be reused easily. I guess you need to implement your own aggregate classes.

Implementing your own classes has another benefit to decrease our dependency on Flink. We'd also like to support other systems in the future (we're currently implementing Hive support for example) and if we rely too much on Flink it would be a headache to identify what should be exported from Flink.